source/language/pulses.rst

```diff
@@ -103,3 +102,62 @@ circuit. Most of the time symbols in the ``defcal`` table will also have
 corresponding ``gate`` definitions. However, if a user provides a ``defcal`` for a symbol
 without a corresponding ``gate``, then we treat such operations like the ``opaque`` gates
 of prior versions of OpenQASM.
+
+Inline calibration blocks
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As calibration grammars may require the ability to insert top-level configuration information, perform program setup, or make inline calls
+to calibration-level instructions, OpenQASM supports the ability to declare a ``cal`` block. Within the ``cal`` block the
+semantics of the selected calibration grammar are valid. The ``cal`` block is of the same scope level as the enclosing block. The calibration
+grammar implementer may therefore choose to allow referencing or modifying values outside of the ``cal`` block but within the containing scope.
+Values declared within the ``cal`` block are only referenceable from other ``cal`` blocks or ``defcal``s that may observe that scope as defined
+by the calibration grammar implementer.
+values may not leak back to the enclosing blocks scope.
+In practice, calibration grammars such as OpenPulse may apply
+a global scope to all identifiers in order to declare values shared across all ``defcal`` calls thereby linking them together.
+
+.. code-block:: c
+
+   OPENQASM 3;
+   defcalgrammar "openpulse";
+
+
+   cal {
+      // declare global channel
+      getchannel d0 = getchannel("drive", $0);
+      // declare global frame
+      frame d0f = newframe(5.0e9, 0.0);
+
+   }
+
+   defcal x $0 {
+      waveform xp = gaussian(1.0, 160t, 40dt);
+      // References channel and frame declared in top-level cal block
+      play(d0, xp, d0f);
+   }
+
+
+Restrictions on defcal bodies
```

Comment:
Proposal: do not allow control flow in OpenPulse. 

Due to the fact we decided to have two grammars, it's a good practise to better distinct the two grammars. Thus, we should prefer some operations in one over the other (otherwise we might as well just have one grammar). In this case, IMO control flow should be limited in OpenQASM grammar. AFAIK, reset is the only supporting example of introducing control flow in OpenPulse and the `defcal` for reset can be replaced by a subroutine in OpenQASM:
```
def reset (bit res) qubit a {
        if (res == 1){
               x a;
        }
}
```
In this case, we also don't have to make the two branches have same lengths.

With disallowing controlflow in OpenPulse, we fully utilize the advantage of having two grammar and compiler implementation will be simplified.

Reply:
I think this is a good solution, however, the the `reset` instruction is a builtin keyword in both OpenQASM2/3. Perhaps we need to define a workaround such as defining that the implementation will link `reset` keyword to a subroutine labeled `__reset__`? Eg.,

```
def __reset__ (bit res) qubit a {
        if (res == 1){
               x a;
        }
}

qubit q;
// Uses new reset
reset q;
```

Reply:
Opened up an issue here for further discussion: https://github.com/Qiskit/openqasm/issues/256 - there's potential complications with https://github.com/Qiskit/openqasm/issues/244